### ISSUE TITLE
'ttnn.where' erroneously propagates NaNs

### ISSUE THREAD

**Issue by sgligorijevicTT** [tenstorrent/tt-metal#22308]
### Describe the bug

Running 'ttnn.where' where one tensor has all NaNs but the condition tensor is such that no values from that tensor should be picked, still results in NaN values on the output.

### Steps to reproduce the issue

1. Run the following code:
```
import torch
import ttnn


def test_ttnn_where():
    C = torch.ones(1, 4, 1, dtype=torch.float32)
    T = torch.randn(1, 4, 768, dtype=torch.float32)
    F = torch.ones(1, 4, 768, dtype=torch.float32) * float('nan')
    golden = torch.where(C != 0, T, F)

    with ttnn.manage_device(0) as dev:
        ttnn_C = ttnn.from_torch(C, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
        ttnn_T = ttnn.from_torch(T, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
        ttnn_F = ttnn.from_torch(F, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
        ttnn_result = ttnn.where(ttnn_C, ttnn_T, ttnn_F)
        result = ttnn.to_torch(ttnn_result)
        print(result)
    print(golden)

    
if __name__ == "__main__":
    test_ttnn_where()
```
2. Observe NaNs being printed

### Expected behavior

Where operation returns the values from the other tensor.

FYI I have already filed another bug also regarding WhereOp, https://github.com/tenstorrent/tt-metal/issues/21439
For a bit more extra info: While debugging model accuracy trough forge a few days ago, a WhereOp like in the unit test appeared to introduce mild inaccuracy by flushing small values to 0 (very similar to https://github.com/tenstorrent/tt-metal/issues/18955), that prompted me to try writing a repro in metal today. A commit that changes the behavior of ttnn.where must have landed sometime inbetween then and now.

### Please complete the following environment information

Ubuntu 22.04, default IRD image
Metal commit: fec47ca3636623b325257d5d6db357003644f550
N150 board

**Comment by mrakitaTT** [comment]
Seems like a recent regression, raising priority since this causes NaNs in some of our models.

**Comment by smehtaTT** [comment]
+ @cmaryanTT

**Comment by cmaryanTT** [comment]
@KalaivaniMCW was this a binary_ng change in the behaviour of where? Can your team please investigate this test case.

**Comment by KalaivaniMCW** [comment]
Looking into this.

**Comment by KalaivaniMCW** [comment]
@cmaryanTT  This seems to be happening only on binary_ng. 
I suspect it is because `ttnn.where` as of now depends on add and mul operations. 
We're still waiting on https://github.com/tenstorrent/tt-metal/issues/19975
It is strange that add & mul behaves differently in ng. will debug more & update.

[where_bh.txt](https://github.com/user-attachments/files/20393723/where_bh.txt)
[where_BH_ng.txt](https://github.com/user-attachments/files/20393725/where_BH_ng.txt)
[where_wh_ng.txt](https://github.com/user-attachments/files/20393724/where_wh_ng.txt)
[where_wh.txt](https://github.com/user-attachments/files/20393726/where_wh.txt)

**Comment by KalaivaniMCW** [comment]
when using multiply-LLK i.e add_binary_tile(), we get 0 * NaN = NaN  
In binary_ng, we only use this LLK for all subtile cases and in legacy binary we use this LLK for non-bcast cases. 

when using bcast kernels i.e BcastOpMath::MUL with BcastOpDim::W, we get 0 * NaN = 0 
We use BcastOpMath in legacy binary for broadcast on H/W/HW dims.

**legacy binary w/o bcast**

```
                     Op | INFO     |  ****** gtz
ttnn.Tensor([[[ 1.00000,  1.00000,  1.00000,  1.00000],
              [ 1.00000,  1.00000,  1.00000,  1.00000],
              [ 1.00000,  1.00000,  1.00000,  1.00000],
              [ 1.00000,  1.00000,  1.00000,  1.00000]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** lez
ttnn.Tensor([[[ 0.00000,  0.00000,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  0.00000,  0.00000]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** using legacy
                     Op | INFO     |  ****** t2 = gtz * T
ttnn.Tensor([[[-0.40193, -0.99981, -0.39809, -0.44792],
              [-1.25204,  2.07705,  0.05912, -1.58995],
              [ 0.96770, -0.29552, -1.39392,  0.45935],
              [-0.27755, -0.16190, -2.26058, -0.15504]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** using legacy
                     Op | INFO     |  ****** t1 = lez * F => 0 * Nan
ttnn.Tensor([[[     nan,      nan,      nan,      nan],
              [     nan,      nan,      nan,      nan],
              [     nan,      nan,      nan,      nan],
              [     nan,      nan,      nan,      nan]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** using legacy
                     Op | INFO     |  ****** out
ttnn.Tensor([[[     nan,      nan,      nan,      nan],
              [     nan,      nan,      nan,      nan],
              [     nan,      nan,      nan,      nan],
              [     nan,      nan,      nan,      nan]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
tt result
TorchTensor([[[nan, nan, nan, nan],
              [nan, nan, nan, nan],
              [nan, nan, nan, nan],
              [nan, nan, nan, nan]]])
tor result
tensor([[[-0.40193, -0.99981, -0.39809, -0.44792],
         [-1.25204,  2.07705,  0.05912, -1.58995],
         [ 0.96770, -0.29552, -1.39392,  0.45935],
         [-0.27755, -0.16190, -2.26058, -0.15504]]])
```

legacy binary w bcast
```

                     Op | INFO     |  ****** gtz
ttnn.Tensor([[[ 1.00000],
              [ 1.00000],
              [ 1.00000],
              [ 1.00000]]], shape=Shape([1, 4, 1]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** lez
ttnn.Tensor([[[ 0.00000],
              [ 0.00000],
              [ 0.00000],
              [ 0.00000]]], shape=Shape([1, 4, 1]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** using legacy
                     Op | INFO     |  ****** using new bcast_w kernel BcastOpMath::MUL
                     Op | INFO     |  ****** t2 = gtz * T
ttnn.Tensor([[[ 1.70703,  0.24390, -0.22314,  0.27295],
              [-0.84668, -0.91602, -1.79688,  1.12109],
              [-1.81641, -0.86523,  1.27148,  0.07996],
              [ 0.66992,  0.87402,  0.50000,  0.76562]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** using legacy
                     Op | INFO     |  ****** using new bcast_w kernel BcastOpMath::MUL
                     Op | INFO     |  ****** t1 = lez * F => 0 * NaN
ttnn.Tensor([[[ 0.00000,  0.00000,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  0.00000,  0.00000]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
                     Op | INFO     |  ****** using legacy
                     Op | INFO     |  ****** out
ttnn.Tensor([[[ 1.70703,  0.24390, -0.22314,  0.27295],
              [-0.84668, -0.91602, -1.79688,  1.12109],
              [-1.81641, -0.86523,  1.27148,  0.07996],
              [ 0.66992,  0.87402,  0.50000,  0.76562]]], shape=Shape([1, 4, 4]), dtype=DataType::FLOAT32, layout=Layout::TILE)
tt result
TorchTensor([[[ 1.70703,  0.24390, -0.22314,  0.27295],
              [-0.84668, -0.91602, -1.79688,  1.12109],
              [-1.81641, -0.86523,  1.27148,  0.07996],
              [ 0.66992,  0.87402,  0.50000,  0.76562]]])
tor result
tensor([[[ 1.70705,  0.24410, -0.22319,  0.27309],
         [-0.84747, -0.91618, -1.79690,  1.12291],
         [-1.81733, -0.86543,  1.27332,  0.08003],
         [ 0.67012,  0.87416,  0.50048,  0.76650]]])
```

**Comment by KalaivaniMCW** [comment]
According to torch results & IEEE standard, 0 * NaN = NaN is the correct behaviour. 
So we have to look at https://github.com/tenstorrent/tt-metal/issues/19975 without involving add/mul in ttnn.where

**Comment by mrakitaTT** [comment]
Thank you for the investigation @KalaivaniMCW!

It seems like there is also some development in https://github.com/tenstorrent/tt-metal/issues/20868 to create a non-composite version of `ttnn.where` without llk changes, but I'm not sure if that will fix this issue because based on [this](https://github.com/tenstorrent/tt-metal/issues/20868#issuecomment-2893683673) it will still use multiplication, what do you think?

**Comment by KalaivaniMCW** [comment]
@mrakitaTT  Yes, as long as we use the multiply LLK, we'll still see this behaviour. 
That is why we postponed writing a device op for ttnn.where since we don't want to depend on add/mul and wait til we can do the conditional check directly on SFPU with a ternary_sfpu LLK. 
But if we still want to use multiply, unless we handle it at multiply LLK by adding a conditional check to return 0's for NaN as well .. the device op approach will also face the same problem.

**Comment by mrakitaTT** [comment]
Understood, thank you @KalaivaniMCW! I will check with LLK owners on ETA for ternary `where` LLK support.

**Comment by KalaivaniMCW** [comment]
@cmaryanTT @patrickroberts 
While we wait for `ttnn.where` LLK, shall we add legacy flag to the `where` op ?

**Comment by cmaryanTT** [comment]
@dchenTT Can we get a list of OPs that are reliant on ttnn.where as part of their composite construction, so we know which OPs will have legacy broadcasting support or less performant broadcasting. Within eltwise, for documentation purposes; and outside of eltiwise just for curiosity.

@patrickroberts can you please get a PR going for Kalaivani's suggestion of using legacy mode on the binary OPs within the ttnn.where implementation. This can be in parallel with on-by-default.

**Comment by dchenTT** [comment]
Ttnn::where Usage

Binary Composite op

_xlogy
_nextafter
_isclose
_atan2
_scatter
Div
Reminder
Fmod
Floor
Power
BackwardAtan2
BackwardXlogy
BackwardDiv


Unary Composite Op

_acosh
_asinh
_atanh
_logit
_lgamma
Trunc
Clamp
Threshold
Rdiv

Ternary Composite Op

_addcdiv

External

ttnn/cpp/ttnn/operations/experimental/reduction/argmax/argmax.cpp
tt-train/sources/ttml/ops/scaled_dot_product_attention.cpp
models/experimental/grok/demo/demo.py
models/demos/t3000/mixtral8x7b/demo/demo.py

**Comment by cmaryanTT** [comment]
@patrickroberts to provide analysis of past and current behaviour and the test issue he identified

**Comment by patrickroberts** [comment]
I've been looking into this and we can't just unconditionally use the legacy implementation of multiply and add in `where_impl` because it doesn't handle broadcasting correctly in cases where binary_ng does fix.

I was able to reproduce the behavior described where binary_ng propagates NaN when it's not selected, but I'm still trying to figure out a suitable solution to this. I'm considering exposing the `use_legacy` parameter to `where` signature so that the user can decide whether they want binary_ng to handle all broadcasting edge-cases properly, or legacy to handle NaN propagation correctly, but right now we don't have a straightforward way to provide both, at least not to my knowledge.

**Comment by patrickroberts** [comment]
Okay so looking into this further, here's a test I'm using to verify the behavior I'm seeing:

```py
def assert_where_with_pcc(torch_input_tensor, torch_input1, torch_input2, device, pcc=0.9999):
    def from_torch_if_tensor(x):
        if not isinstance(x, torch.Tensor):
            return x

        return ttnn.from_torch(x, layout=ttnn.TILE_LAYOUT, device=device)

    input_tensor, input1, input2 = (
        from_torch_if_tensor(arg) for arg in (torch_input_tensor, torch_input1, torch_input2)
    )
    golden_fn = ttnn.get_golden_function(ttnn.where)
    torch_output_tensor = golden_fn(torch_input_tensor > 0, torch_input1, torch_input2)
    output_tensor = ttnn.where(input_tensor, input1, input2)
    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_output_tensor, output_tensor, pcc)


@pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
def test_where_nans(device, dtype):
    torch.manual_seed(0)

    C = torch.ones(1, 4, 1, dtype=dtype)
    T = torch.randn(1, 4, 768, dtype=dtype)
    F = torch.full((1, 4, 768), float("nan"), dtype=dtype)

    assert_where_with_pcc(C, T, F, device)
```
And when I run `pytest 'tests/ttnn/unit_tests/operations/eltwise/test_ternary.py::test_where_nans[dtype=torch.float32]'` using the legacy implementation for `multiply` and `add`, I get the "expected" behavior.

The problem is, this actually depends on a quirk of the legacy implementation, which falls back to the FPU implementation of multiply (`nan * 0 is 0`) when subtile broadcasting is required. If I change the test to the following:

```py
@pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
def test_where_nans(device, dtype):
    torch.manual_seed(0)

    C = torch.ones(1, 4, 768, dtype=dtype)
    T = torch.randn(1, 4, 768, dtype=dtype)
    F = torch.full((1, 4, 768), float("nan"), dtype=dtype)

    assert_where_with_pcc(C, T, F, device)
```

the legacy implementation now exhibits the same unexpected behavior that binary_ng does because it uses the SFPU implementation of multiplication (`nan * 0 is nan`).

**Comment by skhatriTT** [comment]
@patrickroberts do we have an ETA on when it will be fixed?

**Comment by ajakovljevicTT** [comment]
@patrickroberts it seems that this bug also affects 2/3 of our models on the tt-xla frontend (https://github.com/tenstorrent/tt-mlir/issues/3995). Do you have any more information on the progress? Thanks!

**Comment by cmaryanTT** [comment]
We are going to discuss our path for completing this tomorrow morning. I'll have a better ETA then.

I'd like to use our proposed long term solution for the where problem; but need to confirm how far we are from being able to complete that.

**Comment by cmaryanTT** [comment]
ETA for a fixed where is ~2 weeks.

**Comment by KalaivaniMCW** [comment]
This NaN behaviour is now fixed for non-bcast inputs.
support for broadcast cases is in progress. 

<img width="900" height="821" alt="Image" src="https://github.com/user-attachments/assets/7629ff00-63a6-4c74-a7b1-df266214d4d1" />

**Comment by KalaivaniMCW** [comment]
#26322 is on the way for col-bcast support

**Comment by fahmedchanTT** [comment]
This should close after https://github.com/tenstorrent/tt-metal/pull/26322 merges, right @KalaivaniMCW

**Comment by KalaivaniMCW** [comment]
Yes @fahmedchanTT

**Comment by KalaivaniMCW** [comment]
https://github.com/tenstorrent/tt-metal/pull/26322 merged. 
W-dim bcast is now available for where device op.
